compiler/rustc_infer/src/infer/higher_ranked/mod.rs

```diff
@@ -57,6 +57,40 @@ impl<'a, 'tcx> CombineFields<'a, 'tcx> {
         // placeholders which **must not** be named afterwards.
         Ok(())
     }
+
+    #[instrument(skip(self), level = "debug")]
+    pub fn higher_ranked_equate<T>(
+        &mut self,
+        a: Binder<'tcx, T>,
+        b: Binder<'tcx, T>,
+        a_is_expected: bool,
+    ) -> RelateResult<'tcx, ()>
+    where
+        T: Relate<'tcx>,
+    {
+        let span = self.trace.cause.span;
+        // First, we instantiate each bound region in the supertype with a
+        // fresh placeholder region. Note that this automatically creates
+        // a new universe if needed.
+        let b_prime = self.infcx.instantiate_binder_with_placeholders(b);
+
+        // Next, we instantiate each bound region in the subtype
+        // with a fresh region variable. These region variables --
+        // but no other preexisting region variables -- can name
+        // the placeholders.
+        let a_prime = self.infcx.instantiate_binder_with_fresh_vars(span, HigherRankedType, a);
+
+        debug!("a_prime={:?}", a_prime);
+        debug!("b_prime={:?}", b_prime);
+
+        // Compare types now that bound regions have been replaced.
+        let result = self.equate(a_is_expected).relate(a_prime, b_prime)?;
```

Comment:
why do we not also relate a instantiated with placeholders with b instantiated with infer vars? this `higher_ranked_equate` has the same behaviour as relating binders with covariance

Reply:
I think you didn't note that at the call site we do `higher_ranked_equate(a, b)` and `higher_ranked_equate(b, a)`. Anyway, I've inlined the code in Equate and things will be simpler now.